src/main.c

```diff
@@ -0,0 +1,298 @@
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <os_io_seproxyhal.h>
+#include "glyphs.h"
+#include "harmony.h"
+#include "ux.h"
+
+
+commandContext global;
+ux_state_t ux;
+
+static const ux_menu_entry_t menu_main[];
+
+static const ux_menu_entry_t menu_about[] = {
+	{
+		.menu     = NULL,       // another menu entry, displayed when this item is "entered"
+		.callback = NULL,       // a function that takes a userid, called when this item is entered
+		.userid   = 0,          // a custom identifier, helpful for implementing custom menu behavior
+		.icon     = NULL,       // the glyph displayed next to the item text
+		.line1    = "Version",  // the first line of text
+		.line2    = APPVERSION, // the second line of text; if NULL, line1 will be vertically centered
+		.text_x   = 0,          // the x offset of the lines of text; only used if non-zero
+		.icon_x   = 0,          // the x offset of the icon; only used if non-zero
+	},
+
+	{menu_main, NULL, 0, &C_icon_back, "Back", NULL, 61, 40},
+	UX_MENU_END,
+};
+
+static const ux_menu_entry_t menu_main[] = {
+	{NULL, NULL, 0, NULL, "Waiting for", "commands...", 0, 0},
+	{menu_about, NULL, 0, NULL, "About", NULL, 0, 0},
+	{NULL, os_sched_exit, 0, &C_icon_dashboard, "Quit app", NULL, 50, 29},
+	UX_MENU_END,
+};
+
+
+void ui_idle(void) {
+	UX_MENU_DISPLAY(0, menu_main, NULL);
+}
+
+// io_exchange_with_code is a helper function for sending response APDUs from
+// button handlers. Note that the IO_RETURN_AFTER_TX flag is set. 'tx' is the
+// conventional name for the size of the response APDU, i.e. the write-offset
+// within G_io_apdu_buffer.
+void io_exchange_with_code(uint16_t code, uint16_t tx) {
+	G_io_apdu_buffer[tx++] = code >> 8;
+	G_io_apdu_buffer[tx++] = code & 0xFF;
+	io_exchange(CHANNEL_APDU | IO_RETURN_AFTER_TX, tx);
+}
+
+// The APDU protocol uses a single-byte instruction code (INS) to specify
+// which command should be executed. We'll use this code to dispatch on a
+// table of function pointers.
+#define INS_GET_VERSION    0x01
+#define INS_GET_PUBLIC_KEY 0x02
+
+//disable SIGN HASH for security reason
+//#define INS_SIGN_HASH      0x04
+#define INS_GET_TXN_HASH   0x08
+
+// This is the function signature for a command handler. 'flags' and 'tx' are
+// out-parameters that will control the behavior of the next io_exchange call
+// in one_main. It's common to set *flags |= IO_ASYNC_REPLY, but tx is
+// typically unused unless the handler is immediately sending a response APDU.
+typedef void handler_fn_t(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLength, volatile unsigned int *flags, volatile unsigned int *tx);
+
+handler_fn_t handleGetVersion;
+handler_fn_t handleGetPublicKey;
+//handler_fn_t handleSignHash;
+handler_fn_t handleSignTx;
+
+static handler_fn_t* lookupHandler(uint8_t ins) {
+	switch (ins) {
+	case INS_GET_VERSION:    return handleGetVersion;
+	case INS_GET_PUBLIC_KEY: return handleGetPublicKey;
+	//case INS_SIGN_HASH:      return handleSignHash;
```

Comment:
why signHash disabled

Reply:
Disabled just for security reason. As there will be no display of tx related information in ledger Nano S. if anyone hacked the client side app , they can send any hash (from any tx) to ledger Nano S to sign.